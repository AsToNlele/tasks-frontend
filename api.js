import axios from 'axios';
import {
  TASKS_API_ROOT,
  AVAILABLE_TASKS_ROOT,
  EXECUTED_TASKS_ROOT,
  SYSTEMS_ROOT,
} from './src/constants';

const getTasks = async (path) => {
  let response;

  const request = await axios
    .get(TASKS_API_ROOT.concat(path))
    .catch(function (error) {
      return error;
    });

  if (request.status === 200) {
    response = request.data;
  } else {
    response = request;
  }

  return response;
};

export const fetchAvailableTasks = () => {
  return getTasks(AVAILABLE_TASKS_ROOT);
};

export const fetchAvailableTask = (slug) => {
  return getTasks(AVAILABLE_TASKS_ROOT.concat(`/${slug}`));
};

export const fetchExecutedTasks = () => {
  return getTasks(EXECUTED_TASKS_ROOT);
};

export const fetchExecutedTask = (path) => {
  return getTasks(EXECUTED_TASKS_ROOT.concat(path));
};

export const fetchExecutedTaskJobs = (id) => {
  let idPath = `/${id}`;
  return fetchExecutedTask(idPath.concat('/jobs'));
};

export const fetchSystems = (path) => {
  return getTasks(SYSTEMS_ROOT.concat(path));
};
