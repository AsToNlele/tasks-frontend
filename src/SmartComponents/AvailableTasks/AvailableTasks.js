import React from 'react';

const AvailableTasks = () => {
  return <div aria-label="available-tasks">Available tasks</div>;
};

export default AvailableTasks;
