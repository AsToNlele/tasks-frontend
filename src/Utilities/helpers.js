import React from 'react';
import moment from 'moment';
import {
  Skeleton,
  SkeletonSize,
} from '@redhat-cloud-services/frontend-components/Skeleton';

const renderRunning = (time) => {
  return time === 'null' || !time ? 'Running' : false;
};

export const renderRunDateTime = (time) => {
  if (time === 'loading') {
    return <Skeleton size={SkeletonSize.md} />;
  } else {
    return (
      renderRunning(time) || moment.utc(time).format('DD MMM YYYY, HH:mm UTC')
    );
  }
};

export const getTimeDiff = ([start, end]) => {
  if (start === 'loading') {
    return <Skeleton size={SkeletonSize.md} />;
  } else {
    return (
      renderRunning(end) ||
      `${renderRunDateTime(end)} (${moment
        .duration(
          moment(renderRunDateTime(end), 'DD MMM YYYY, HH:mm').diff(
            moment(renderRunDateTime(start), 'DD MMM YYYY HH:mm')
          )
        )
        .asMinutes()} min)`
    );
  }
};

export const uniq = (collection) => [...new Set(collection)];

/*eslint-disable react/display-name*/
export const renderColumnComponent =
  (Component, props) => (_data, _id, entity) =>
    <Component {...entity} {...props} />;
/*eslint-enable react/display-name*/

export const orderArrayByProp = (property, objects, direction) =>
  objects.sort((a, b) => {
    if (direction === 'asc') {
      if (typeof a[property] === 'number') {
        return a[property] - b[property];
      } else {
        return String(a[property]).localeCompare(String(b[property]));
      }
    } else {
      if (typeof a[property] === 'number') {
        return -a[property] + b[property];
      } else {
        return -String(a[property]).localeCompare(String(b[property]));
      }
    }
  });

export const getProperty = (obj, path, fallback) => {
  const parts = path.split('.');
  const key = parts.shift();
  if (typeof obj[key] !== 'undefined') {
    return parts.length > 0
      ? getProperty(obj[key], parts.join('.'), fallback)
      : obj[key];
  }

  return fallback;
};

export const camelCase = (string) => {
  return string
    .split(/[-_\W]+/g)
    .map((string) => string.trim())
    .map((string) => string[0].toUpperCase() + string.substring(1))
    .join('');
};
