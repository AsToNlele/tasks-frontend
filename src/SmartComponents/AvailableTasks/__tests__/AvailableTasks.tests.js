import React from 'react';
import { act, render, screen, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';

import AvailableTasks from '../AvailableTasks';
import {
  availableTasksTableError,
  availableTasksTableItems,
} from '../../../Utilities/hooks/useTableTools/Components/__tests__/TasksTable.fixtures';
import { dispatchNotification } from '../../../Utilities/Dispatcher';
import { fetchAvailableTasks } from '../../../../api';

jest.mock('../../../../api');
jest.mock('../../../Utilities/Dispatcher');

describe('AvailableTasks', () => {
  let mockStore = configureStore();
  let props;

  beforeEach(() => {
    props = {
      openTaskModal: jest.fn(),
    };
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it.skip('should render correctly', async () => {
    fetchAvailableTasks.mockImplementation(
      async () => availableTasksTableItems
    );

    const { asFragment } = act(async () => {
      render(
        <MemoryRouter keyLength={0}>
          <AvailableTasks {...props} />
        </MemoryRouter>
      );
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it.skip('should fetch api data and build table', async () => {
    fetchAvailableTasks.mockImplementation(
      async () => availableTasksTableItems
    );

    const store = mockStore(props);
    await act(async () => {
      render(
        <MemoryRouter keyLength={0}>
          <Provider store={store}>
            <AvailableTasks {...props} />
          </Provider>
        </MemoryRouter>
      );
    });

    waitFor(() => {
      expect(
        screen.getByLabelText('available-tasks-table').toBeInTheDocument()
      );
      expect(screen.getByLabelText('taska').toBeInTheDocument());
    });
  });

  it.skip('should render error', async () => {
    const mockedNotification = jest.fn();
    fetchAvailableTasks.mockImplementation(
      async () => availableTasksTableError
    );
    dispatchNotification.mockImplementation(mockedNotification);

    const store = mockStore(props);
    await act(async () => {
      render(
        <MemoryRouter keyLength={0}>
          <Provider store={store}>
            <AvailableTasks {...props} />
          </Provider>
        </MemoryRouter>
      );
    });

    waitFor(() => {
      //const thing = screen.getByLabelText('taska');
      expect(screen.getByLabelText('taska')).toBeTruthy();
    });
  });
});
