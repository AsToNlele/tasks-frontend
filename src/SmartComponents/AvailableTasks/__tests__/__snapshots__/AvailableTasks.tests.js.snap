// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvailableTasks should render correctly 1`] = `
<DocumentFragment>
  <div
    aria-label="available-tasks"
  >
    <article
      class="pf-c-card"
      data-ouia-component-id="OUIA-Generated-Card-1"
      data-ouia-component-type="PF4/Card"
      data-ouia-safe="true"
      id=""
    >
      <div
        class="pf-c-card__title"
      >
        <div
          class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
        >
          <span
            class="pf-u-screen-reader"
          />
        </div>
      </div>
      <div
        class="pf-c-card__body"
      >
        <div
          class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
        >
          <span
            class="pf-u-screen-reader"
          />
        </div>
      </div>
      <div
        class="pf-c-card__footer"
      >
        <div
          class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
        >
          <span
            class="pf-u-screen-reader"
          />
        </div>
      </div>
    </article>
    <br />
    <article
      class="pf-c-card"
      data-ouia-component-id="OUIA-Generated-Card-2"
      data-ouia-component-type="PF4/Card"
      data-ouia-safe="true"
      id=""
    >
      <div
        class="pf-c-card__title"
      >
        <div
          class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
        >
          <span
            class="pf-u-screen-reader"
          />
        </div>
      </div>
      <div
        class="pf-c-card__body"
      >
        <div
          class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
        >
          <span
            class="pf-u-screen-reader"
          />
        </div>
      </div>
      <div
        class="pf-c-card__footer"
      >
        <div
          class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
        >
          <span
            class="pf-u-screen-reader"
          />
        </div>
      </div>
    </article>
    <br />
  </div>
</DocumentFragment>
`;
